File: java-code/src/exceptions/Exceptions.java
package exceptions;

import classes.Manager;
import classes.base.Employee;

public class Exceptions {
  public static int validateInt(String string) throws NumberFormatException {
    try {
      int number = Integer.parseInt(string);

      if (number < 0) {
        throw new NumberFormatException();
      }
      return number;
    } catch (NumberFormatException e) {
      System.err.println("Invalid Number inputted - Please input a valid positive number - " +
              "ex/ 5");
      throw new NumberFormatException("Invalid number inputted: " + e);
    }
  }
  public static double validateDouble(String string) throws NumberFormatException {
    try {
      double number = Double.parseDouble(string);

      if (number < 0.0) {
        throw new NumberFormatException();
      }
      return number;
    } catch (NumberFormatException e) {
      System.err.println("Invalid Number inputted - Please input a valid positive number - " +
              "ex/ 5.6");
      throw new NumberFormatException("Invalid number inputted: " + e);
    }
  }

  public static String validateString(String string) throws RuntimeException {
    try {
      if (string == null || string.isEmpty()) {
        throw new RuntimeException();
      }
      return string;
    } catch (RuntimeException e) {
      System.err.println("One or more string is empty..");
      throw new RuntimeException("Empty String " + e);
    }
  }

  public static char validateYesNo(String string) throws RuntimeException {
    try {
      if (!string.equalsIgnoreCase("y") || !string.equalsIgnoreCase("n")) {
        throw new RuntimeException();
      }
      return string.charAt(0);
    } catch (RuntimeException e) {
      System.err.println("Invalid input detected");
      throw new RuntimeException("Invalid input " + e);
    }
  }

  public static void validateEmployee(Employee employee) throws NullPointerException {
    try {
      if (employee == null) {
        throw new NullPointerException();
      }
    } catch (NullPointerException e) {
      System.err.println("Employee does not exist");
      throw new NullPointerException("Employee does not exist " + e);
    }
  }

  public static void validateManager(Manager manager) throws NullPointerException {
    try {
      if (manager == null) {
        throw new NullPointerException();
      }
    } catch (NullPointerException e) {
      System.err.println("Manager does not exist");
      throw new NullPointerException("Manager does not exist " + e);
    }
  }
}

File: c#-code/src/exceptions/exceptions.cs
namespace APLCAssignment3.Exceptions
{
    using APLCAssignment3.Classes; 
    using APLCAssignment3.Classes.Base; 
    using System;

    public static class Validation
    {
        public static int ValidateInt(string input)
        {
            try
            {
                int number = int.Parse(input);

                if (number < 0)
                {
                    throw new ArgumentOutOfRangeException(nameof(input), "Must be a positive number.");
                }
                return number;
            }
            catch (FormatException e)
            {
                Console.Error.WriteLine("Invalid Number inputted - Please input a valid positive number - ex/ 5");
                throw new FormatException($"Invalid number inputted: {e.Message}", e); 
            }
            catch (ArgumentOutOfRangeException e)
            {
                Console.Error.WriteLine("Invalid Number inputted - Please input a valid positive number - ex/ 5");
                throw new ArgumentOutOfRangeException($"Invalid number inputted: {e.Message}", e);
            }
        }

        public static double ValidateDouble(string input)
        {
            try
            {
                double number = double.Parse(input);

                if (number < 0.0)
                {
                    throw new ArgumentOutOfRangeException(nameof(input), "Must be a positive number.");
                }
                return number;
            }
            catch (FormatException e)
            {
                Console.Error.WriteLine("Invalid Number inputted - Please input a valid positive number - ex/ 5.6");
                throw new FormatException($"Invalid number inputted: {e.Message}", e);
            }
            catch (ArgumentOutOfRangeException e)
            {
                Console.Error.WriteLine("Invalid Number inputted - Please input a valid positive number - ex/ 5.6");
                throw new ArgumentOutOfRangeException($"Invalid number inputted: {e.Message}", e);
            }
        }

        public static string ValidateString(string input) 
        {
            try
            {
                if (string.IsNullOrEmpty(input))
                {
                    throw new ArgumentException("String cannot be null or empty."); 
                }
                return input;
            }
            catch (ArgumentException e)
            {
                Console.Error.WriteLine("One or more string is empty..");
                throw new ArgumentException($"Empty String: {e.Message}", e);
            }
        }

        public static char ValidateYesNo(string input) 
        {
            try
            {
                if (!input.Equals("y", StringComparison.OrdinalIgnoreCase) && 
                    !input.Equals("n", StringComparison.OrdinalIgnoreCase))
                {
                    throw new ArgumentException("Input must be 'y' or 'n'.");
                }
                return input[0]; 
            }
            catch (ArgumentException e)
            {
                Console.Error.WriteLine("Invalid input detected");
                throw new ArgumentException($"Invalid input: {e.Message}", e);
            }
        }

        public static void ValidateEmployee(Employee employee) 
        {
            try
            {
                if (employee == null)
                {
                    throw new ArgumentNullException(nameof(employee));
                }
            }
            catch (ArgumentNullException e)
            {
                Console.Error.WriteLine("Employee does not exist");
                throw new ArgumentNullException($"Employee does not exist: {e.Message}", e);
            }
        }

        public static void ValidateManager(Manager manager) 
        {
            try
            {
                if (manager == null)
                {
                    throw new ArgumentNullException(nameof(manager));
                }
            }
            catch (ArgumentNullException e)
            {
                Console.Error.WriteLine("Manager does not exist");
                throw new ArgumentNullException($"Manager does not exist: {e.Message}", e);
            }
        }
    }
}